soc: arm: nordic_nrf: nrf53: Add Synchronized RTC

nordic-krch opened the pull request:
The goal of this PR is to establish a mechanism which allows to get the offset between system clock on network and application core (assuming they both run from RTC). Offset can be applied to a function which returns timestamp for logging allowing logs synchronization from both cores (currently on independent outputs).

Add a module which is responsible for getting offset between RTCs
used for system clock on NET and APP cores. After getting an offset
between NET and APP clocks, it can be used for logging timestamping
on NET core to ensure that timestamping is in sync on both cores.
    
Synchronization is done using PPI, IPM task and events and RTC
capture feature.

Contains also extension to `nrf_rtc_timer` by adding function for getting CAPTURE task address `z_nrf_rtc_timer_capture_task_address_get()` (not present on all nRF platforms).

Additionally added sample that shows results of RTC synchronization for nrf53 `samples/boards/nrf/nrf53_sync_rtc`

Fixes #23021

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: arm: nordic_nrf: nrf53: Add Synchronized RTC

Add a module which is responsible for getting offset between RTCs
used for system clock on NET and APP cores. After getting an offset
between NET and APP clocks, it can be used for logging timestamping
on NET core to ensure that timestamping is in sync on both cores.

Synchronization is done using PPI, IPM task and events and RTC
capture feature.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
samples: boards: nrf: Add nrf53_sync_rtc sample

Add sample which demonstrates how RTC clocks are synchronized
between application and network core.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on samples/boards/nrf/nrf53_sync_rtc/README.rst:
```diff
@@ -0,0 +1,76 @@
+   :zephyr-app: samples/boards/nrf/nrf53_sync_rtc
+   :board: nrf5340dk_nrf5340_cpuapp
+   :goals: flash
+   :goals: flash --hex-file build/sync_rtc_net-prefix/src/sync_rtc_net-build/zephyr/zephyr.hex
```
@mbolivar not sure if there is a better way to flash `ExternalProject`. In ninja i can type `ninja sync_rtc_net flash` but couldn't find west equivalent.

koffes commented:
Fantastic feature! I will have a look at the presentation as soon as I have time.

nordic-krch committed:
soc: arm: nordic_nrf: nrf53: Add nrf53 series guards to Kconfig.soc

All options in Kconfig.soc are only for nrf53 series. Added file
wide guard instead of option specific options.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on soc/arm/nordic_nrf/nrf53/sync_rtc.c:
```diff
@@ -0,0 +1,244 @@
+		ppi_ipc_to_rtc(channels, true);
+	} else {
+		ppi_rtc_to_ipc(channels, true);
+		sync_cc = z_nrf_rtc_timer_read() + RTC_SYNC_ARBITRARY_OFFSET;
+		z_nrf_rtc_timer_compare_set(channels.ch.rtc, sync_cc,
+					    rtc_cb, (void *)channels.raw);
```
The use of `RTC_SYNC_ARBITRARY_OFFSET` here is a bit confusing, as the value used here has nothing to do with offset between APP and NET, it's only about successful setting of the RTC compare in the near future, right? And I think you should prevent preemption of this part of code.

nordic-krch commented on include/drivers/timer/nrf_rtc_timer.h:
```diff
@@ -7,6 +7,8 @@
 #ifndef ZEPHYR_INCLUDE_DRIVERS_TIMER_NRF_RTC_TIMER_H
 #define ZEPHYR_INCLUDE_DRIVERS_TIMER_NRF_RTC_TIMER_H
 
+#include <sys_clock.h>
```
it is somehow related. `z_nrf_rtc_timer_get_ticks` is using `k_timeout_t` from `sys_clock.h` so it was relying on it being included somewhere else. I can move it to separate commit.

nordic-krch committed:
drivers: timer: nrf_rtc_timer: Add missing include

k_timeout_t is used in the header but sys_clock.h was not included.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: timer: nrf_rtc_timer: Add z_nrf_rtc_timer_capture_task_address_get

Add function for getting address of RTC capture task.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on include/drivers/timer/nrf_rtc_timer.h:
```diff
@@ -46,6 +48,18 @@ uint32_t z_nrf_rtc_timer_read(void);
+ *
+ * @note Not all platforms have CAPTURE task.
+ *
+ * @param chan Channel ID between 0 and CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT.
```
added protection, updated doc

nordic-krch commented on soc/arm/nordic_nrf/nrf53/sync_rtc.c:
```diff
@@ -0,0 +1,244 @@
+#define RTC_SYNC_ARBITRARY_OFFSET 100
+
+static uint32_t sync_cc;
+uint32_t nrf53_sync_offset;
```
added api `z_nrf_rtc_timer_nrf53net_rtc_offset` to `nrf_rtc_timer.h` but with implementation in `sync_rtc.c`

nordic-krch commented:
@anangl comments applied.

nordic-krch commented:
@anangl ping.

nordic-krch commented:
@anangl ping

nordic-krch commented:
Reworked to support use of mbox driver as well.

nordic-krch commented:
@alexsven @thebatfisk @koffes can anyone of you try it if it works ok in your application. There is only simple example running and it would be good to validate it in more complex application.

nordic-krch commented on soc/arm/nordic_nrf/nrf53/sync_rtc.c:
```diff
@@ -0,0 +1,311 @@
+
+	ipm_dev = device_get_binding("IPM_" STRINGIFY(CONFIG_NRF53_SYNC_RTC_IPM_IN));
+	if (ipm_dev == NULL) {
+		return -EIO;
```
done

nordic-krch commented on soc/arm/nordic_nrf/nrf53/sync_rtc.c:
```diff
@@ -0,0 +1,311 @@
+ */
+static void ppi_rtc_to_ipc(union rtc_sync_channels channels, bool setup)
+{
+
```
done

nordic-krch commented on soc/arm/nordic_nrf/nrf53/sync_rtc.c:
```diff
@@ -0,0 +1,311 @@
+ */
+static void ppi_ipc_to_rtc(union rtc_sync_channels channels, bool setup)
+{
+
```
done

nordic-krch commented on include/drivers/timer/nrf_rtc_timer.h:
```diff
@@ -154,6 +154,18 @@ void z_nrf_rtc_timer_abort(int32_t chan);
+ * @retval -ENOSYS if operation is not supported.
+ * @retval -EBUSY if synchronization is not yet completed.
+ */
+int z_nrf_rtc_timer_nrf53net_rtc_offset(void);
```
done
